doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+Examples of tags for data variables are:
+- `Data::Value`
+- `Data::Variance`
+- `Data::String`
+- `Data::Int`
+
+Examples of tags coordinate variables are:
+- `Coord::Tof`
+- `Coord::dSpacing`
+- `Coord::Q`
+- `Coord::SpectrumNumber`
+- `Coord::DetectorId`
+- `Coord::SpectrumPosition`
+- `Coord::DetectorPosition`
+- `Coord::DetectorGrouping`
+- `Coord::RowLabel`
+- `Coord::X`, `Coord::Y`, `Coord::Z`
+- `Coord::Qx`, `Coord::Qy`, `Coord::Qz`
+- coordinates related to experiment environment parameters such as `Coord::Polarization` and `Coord::Temperature`
+- `Coord::Mask`
+
+The tag defines the underlying data type.
+Experiment logs (`API::Run` in Mantid) are probably a special case, see the section on [Attributes](#design-details-attributes) below.
+
+We furthermore define:
+- Coordinate variables do not have a name and are thus identified only by their tag.
+  Since `Dataset` is a set this implies that there cannot be two coordinate variable with the same tag in the same dataset.
+  *Example: Having two variables `Coord::X` representing an X coordinate in a dataset is forbidden, and would indeed not make sense.*
+- Coordinate variables have an implicit unit that cannot be changed (at least apart from its scale) since it is implied by the tag defining the coordinate.
+  *Example: The X coordinate would always have a unit representing a *length*, e.g., `m` or `mm`.*
+- Data variables have a name in addition to their tag.
+  *Example: We can have a data variable holding neutron counts in `Data::Value` fom the "sample" run and a data variable holding counts for the "can" run in the *same* dataset.*
+
+The name of data variables is furthermore used to imply a grouping of variables.
+*Example: In a dataset containing data variables `Data::Value` and `Data::Variance` a matching variable name implies that the uncertainties stored in `Data::Variance` are associated with the data stored in `Data::Value`.*
+
+The data in `Variable` is held by a type-erased handle, i.e., a `Variable` can hold data of any type.
+The data handle also implements a copy-on-write mechanism, which makes `Variable` and `Dataset` cheap to copy and saves memory if only some variables in a dataset are modified.
+
+
+### <a name="overview-operations"></a>Operations
+
+Alongside the new `Dataset` type a series of basic and common operations will be provided.
+A mostly but not entirely complete list of operations to be supported is:
+1. Arithmetic operations: unary `-`, binary `+`, `-`, `*`, `/`
+1. `transpose`
+1. `slice` and `dice`
+1. `concatenate` and `merge`
+1. `sort`
+1. `rebin`
+1. `convert` (unit and/or dimension)
+1. `filter`
+1. Boolean/logical operations for mask operations
+1. `integrate`
+1. Basic statistics operations such as min, max, mean, and standard deviation
+1. Serialization and deserialization, mainly for saving to and loading from disk
+   *Question: Should disk I/O live in a separate module/library?*
```

Comment:
Good point, I should look into it. In principle things are a bit simpler than now, since we are dealing with larger arrays. For other components like `Run` we may reuse the current loading/saving code.

Do you know if there are any requirements for such a Mantid-internal data format (basically the successor of the current output of `SaveNexusProcessed`)?

Reply:
Do you mean formal document of the current requirements for saving? If so then only some very [old documentation](https://github.com/mantidproject/documents/blob/master/Design/Nexus/ISIS%20Nexus/Minimum%20ISIS%20NeXus%20file%20format.doc). I'm not sure what the "modern" NeXus way of doing this is.

Reply:
No, I meant whether we have any restrictions, or whether we are completely free in chosing a format for saving? For example, if we are restricted to Nexus, we will inherit the performance problems we are having with HDF5 (no multi-threaded loading).

Reply:
For our own internal format I don't think we need to particular stick to NeXus/HDF5. If we wrote straight HDF5 could we take advantage of any of the parallel versions?

Reply:
Well, the threaded HDF5 (which is what Mantid is using), is _thread-safe_, but not _concurrent_ (which is one of the main reasons why `LoadEventNexus` is slow). Then there is a MPI-based HDF5 library which is completely different, but I am not sure we want to rely on that...

Reply:
Not for a non-mpi build...

Reply:
The other problem with speed HDF5 is by the way compression: The built-in compression is very slow. There are faster ones like `blosc` which are supported as plugins, but at least here there is resitance to using that since it is basically "non-standard", i.e., may be hard to read in 10 years. I think with `blosc` the single-thread read performance is more adequate.